File: kth_largest-element_in_an_array.py
import random
class Solution:
    def findKthLargest(self, nums, k):
        n = len(nums)
        if n==1:
            return nums[0]
        q = random.randint(0,n-1)
        nums[-1], nums[q] = nums[q], nums[-1]
        pivort = nums[-1]
        i = -1
        for j in range(n-1):
            if nums[j] <= pivort:
                i += 1
                nums[i], nums[j] = nums[j], nums[i]
        nums[i+1], nums[-1] = nums[-1], nums[i+1]
        cnt =n- i-1
        if cnt == k:
            return pivort
        if cnt > k:           
            return self.findKthLargest(nums[n-cnt+1:], k)
        elif cnt < k:
            return self.findKthLargest(nums[:n-cnt],k-cnt)
        
if __name__ == '__main__':
    sol = Solution()
    nums = [2,1]
    print sol.findKthLargest(nums,2),1
    print '-'*19
    nums = [-1,-1]
    print sol.findKthLargest(nums,2),-1
    print '-'*19
    nums = [7,6,5,4,3,2,1]
    print sol.findKthLargest(nums,5),3
    nums = [3,2,3,1,2,4,5,5,6]
    print '-'*19
    print sol.findKthLargest(nums,9),1
    print '-'*19
    nums= [3,2,3,1,2,4,5,5,6,7,7,8,2,3,1,1,1,10,11,5,6,2,4,7,8,5,6]
    print sol.findKthLargest(nums,20),2
    print '-'*19
    nums= [3,2]
    print sol.findKthLargest(nums,2),2
    print '-'*19
    for i in range(1,7):
        nums = [3,2,1,5,6,4]
        print sol.findKthLargest(nums,i)
                
            
        
        

File: Algorithm/CPP/kth_largest_element_in_an_array.cpp
# include <stdio.h>
# include <iostream>
# include <vector>
# include <algorithm>
using namespace std;


class Solution {
public:
    int findKthLargest(vector<int>& nums, int k) {
		int n = nums.size();
		if (n == 1) return nums[0];
		k = n - k;
		//cout << k << " k " << n << " n " << endl;
		return findKthSmallest(nums,0,n-1, k);
	}
	int findKthSmallest(vector<int>& nums, int left, int right, int k)
	{
		int pivort = nums[right];
		int i = left;
		for (int j = left; j < right; j++)
		{
			if (nums[j] <= pivort)
			{
				swap(nums[i++], nums[j]);
			}
			
		}
		swap(nums[i], nums[right]);
		if(i-left == k) return nums[i];
		if (i-left > k) return findKthSmallest(nums,left,i-1, k);
		else return findKthSmallest(nums, i+1, right, k - (i -left+1));
	}
};
int main()
{
	Solution sol;
	int numRows = 5;
	for (auto k : sol.generate(numRows))
	{
		for (auto x : k) cout << x << " ";
		cout << endl;
	}
	system("pause");
	return 0;
}

